src/gc.c

```diff
@@ -62,44 +51,116 @@ GCContext* GCContext_CreateGC(RedisModuleString* keyName, float initialHZ, uint6
+  if (gc->callbacks.kill) {
+    gc->callbacks.kill(gc->gcCtx);
+  }  
+}
+
+static void timerCallback(RedisModuleCtx* ctx, void* data);
+
+static long long getNextPeriod(GCContext* gc) {
+  struct timespec interval = gc->callbacks.getInterval(gc->gcCtx);
+  long long ms = interval.tv_sec * 1000 + interval.tv_nsec / 1000000;  // convert to millisecond
+  return ms;
+}
+
+static RedisModuleTimerID scheduleNext(GCTaskCtx *task) {
+  if (!RedisModule_CreateTimer) return 0;
+
+  long long period = getNextPeriod(task->gc);
+  return RedisModule_CreateTimer(RSDummyContext, period, timerCallback, task);
+}
+
+static void threadCallback(void* data) {
+  GCTaskCtx* task= data;
+  GCContext* gc = task->gc;
+  RedisModuleBlockedClient* bc = task->bClient;
+  RedisModuleCtx* ctx = RSDummyContext;
+
+  if (gc->stopped) {
+    if (bc && bc != DEADBEEF) {
+      RedisModule_ThreadSafeContextLock(ctx);
+      RedisModule_UnblockClient(bc, NULL);
+      RedisModule_ThreadSafeContextUnlock(ctx); 
+    }
+    rm_free(task);
+    return;
   }
 
-  GCContext* gc = privdata;
   int ret = gc->callbacks.periodicCallback(ctx, gc->gcCtx);
 
-  RedisModuleBlockedClient* bClient = BlockClients_pop(&gc->bClients);
-  if (bClient) {
-    RedisModule_UnblockClient(bClient, NULL);
+  RedisModule_ThreadSafeContextLock(ctx);
+  if (bc) { 
+    if (bc != DEADBEEF) {
+      RedisModule_UnblockClient(bc, NULL);
+    }
+    rm_free(task);
+    goto end;
   }
-  if (gc->timer) {
-    // Timer could've been deleted..
-    RMUtilTimer_SetInterval(gc->timer, gc->callbacks.getInterval(gc->gcCtx));
+
+  if (!ret || gc->stopped) {
+    stopGC(gc);
+    rm_free (task);
+    goto end;
   }
-  return ret;
+
+  gc->timerID = scheduleNext(task);
+
+end:
+  RedisModule_ThreadSafeContextUnlock(ctx);
 }
 
-static void GCContext_OnTerm(void* privdata) {
-  GCContext* gc = privdata;
+static void destroyCallback(void* data) {
+  GCContext* gc = data;
+  RedisModuleCtx* ctx = RSDummyContext;
+  assert(gc->stopped == 1);
+
+  RedisModule_ThreadSafeContextLock(ctx);  
   gc->callbacks.onTerm(gc->gcCtx);
   rm_free(gc);
+  RedisModule_ThreadSafeContextUnlock(ctx);
+}
+
+static void timerCallback(RedisModuleCtx* ctx, void* data) {
+  if (RedisModule_AvoidReplicaTraffic && RedisModule_AvoidReplicaTraffic()) {
+    // If slave traffic is not allow it means that there is a state machine running
+    // we do not want to run any GC which might cause a FORK process to start for example).
+    // Its better to just avoid it.
+    GCTaskCtx* task = data;
+    task->gc->timerID = scheduleNext(task);
+    return;
+  }
+  thpool_add_work(gcThreadpool_g, threadCallback, data);
 }
 
 void GCContext_Start(GCContext* gc) {
-  struct timespec interval = gc->callbacks.getInterval(gc->gcCtx);
-  gc->timer = RMUtil_NewPeriodicTimer(GCContext_PeriodicCallback, GCContext_OnTerm, gc, interval);
+  GCTaskCtx* task = GCTaskCreate(gc, NULL);
+  gc->timerID = scheduleNext(task);
+  if (gc->timerID == 0) {
+    rm_free(task);
```

Comment:
Log?

Reply:
This can only happen with debug command. 
RedisModule_CreateTimer always returns a value